lookout/style/typos/analyzer.py

```
@@ -145,17 +155,69 @@ def render_comment_text(self, typo_fix: TypoFix) -> str:
+            old_code_line=file_lines[typo_fix.line_number - 1],
+            confidences=self._normalize_confidences(confidences),
+            zip=zip,
+            enumerate=enumerate,
+            int=int,
+            )
+
+    @staticmethod
+    def _compute_confidence(confidences: Sequence[float]) -> float:
+        confidence = 1
+        for conf in confidences:
+            confidence *= conf
+        return confidence
 
     @staticmethod
-    def _get_comment_confidence(suggestions: Sequence[Candidate]) -> int:
-        invert_confidence = 1
-        for suggestion in suggestions:
-            invert_confidence *= (1 - suggestion.confidence)
-        return int(100 * (1 - invert_confidence))
+    def _normalize_confidences(confidences: Sequence[float]):
+        s = sum(confidences)
+        return [conf / s for conf in confidences]
+
+    def generate_identifier_suggestions(self, suggestions: Mapping[str, Iterable[Candidate]],
+                                        identifier: str, n_candidates: int = 3) \
+            -> Iterator[Tuple[List[str], float]]:
+        """
+        Generate suggestions for identifier and compute probability of suggestion.
+
+        :param suggestions: suggestions - mapping from token to list of candidates.
+        :param identifier: initial identifier.
+        :param n_candidates: number of candidates to generate.
+        :return: generator of (list of suggestion for identifier, probability).
+        """
+        initial_tokens = self.parser.split(identifier)
+        final_sugg = []
+        confidences = [1]
+        suggestion_queue = []
+        for i, t in enumerate(initial_tokens):
+            try:
+                final_sugg.append(suggestions[t][0].token)
+                confidences.append(suggestions[t][0].confidence)
+                for sugg in suggestions[t][1:]:
+                    suggestion_queue.append((i, sugg.token, sugg.confidence))
+            except KeyError:
```

I'd like to have something like:
```python
if t in suggestions:
    ...
else:
    ...
```
instead of try-except. It makes logic move straightforward.

---

A couple of points:
We have quite a lot of places with the same logic to check `key`
[style analyzer](https://github.com/src-d/style-analyzer/search?q=keyerror&unscoped_q=keyerror)
[src-d/ml](https://github.com/src-d/ml/search?q=keyerror&unscoped_q=keyerror)

And several more arguments https://stackoverflow.com/questions/12342822/python-try-except-keyerror-vs-if-has-key?lq=1
TLDR:
it's cheaper to have `exception` in case of rare `KeyErrors` than check `if condition` on every iteration.
To make a proper decision we need to evaluate it on real data to understand if it's worth to switch logic to checking `if condition` at each iteration (frequent `KeyErrors`).

---

I do not like that it hides possible bugs if `suggestions[t]` is empty. 
I do not see performance issues here. 
if you want to keep try-except let's add assert

---

It's handled here https://github.com/src-d/style-analyzer/blob/f00b3eeac4ff0085b4a73c76cc7827b2172d9e87/lookout/style/typos/analyzer.py#L139~L140
How `if key in suggestions` can solve an empty `suggestions` issue?

---

it did not hide the error and you get an exception.
let's just add an assert. I think it is fine.

---

I think that was missed ^

---

ok, outdated.